torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -511,6 +547,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
   GRAPH_DUMP("After remove observers :", g);
 }
 
+Value* ProcessWeightObserverSubgraph::updateInputValueInGraph(
+    Value* v,
+    std::shared_ptr<Graph>& g) {
+  if (remap_values_in_observer_subgraph_.count(v) == 0) {
+    auto new_value = g->block()->addInput();
+    remap_values_in_observer_subgraph_[v] = new_value;
+    new_value->copyMetadata(v);
+    return new_value;
+  } else {
+    return remap_values_in_observer_subgraph_[v];
+  }
+}
+
+void ProcessWeightObserverSubgraph::cloneNodeInGraph(
+    Node* node,
+    std::shared_ptr<Graph>& g) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) { return updateInputValueInGraph(v, g); };
+
+  auto new_node = block->appendNode(g->createClone(node, value_fn));
+  for (size_t i = 0; i < node->outputs().size(); ++i) {
+    auto oo = node->outputs()[i];
+    auto no = new_node->outputs()[i];
+    remap_values_in_observer_subgraph_[oo] = no;
+  }
+}
+void ProcessWeightObserverSubgraph::buildObserverSubgraph(
+    const std::vector<Node*>& weight_subgraph,
+    std::shared_ptr<Graph> dest_graph) {
+  // Build weight subgraph
+  for (auto n : weight_subgraph) {
+    cloneNodeInGraph(n, dest_graph);
+  }
+  LintGraph(dest_graph);
+  // Add last node output value as subgraph output.
+  dest_graph->registerOutput(
+      updateInputValueInGraph(weight_subgraph.back()->output(0), dest_graph));
+  GRAPH_DUMP("New weight observer subgraph: ", dest_graph);
+}
+
+void ProcessWeightObserverSubgraph::findSubgraphNodes(
+    const std::vector<Value*>& inputs,
+    std::vector<Node*>& weight_subgraph) {
+  for (auto v : inputs) {
+    auto n = v->node();
+    if (!hitGraphInput(n->output(0))) {
+      weight_subgraph.push_back(n);
+      findSubgraphNodes(n->inputs().vec(), weight_subgraph);
+    }
+  }
+}
+
+void ProcessWeightObserverSubgraph::extractAndRunWeightObserver(
+    Module& module,
+    Value* self,
+    Node* observer) {
+  Value* observed_weight = observer->output();
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+
+  // If the graph was already visited, return list of relevant nodes directly.
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (graph_to_subgraph_nodes_.count(g)) {
```

Comment:
the subgraph should not be indexed by graph, it should be indexed by value, we'll have one subgraph for each weight value